transaction-payment/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+use payment::OnChargeAssetTransaction;
+
+mod payment;
+// Type aliases used for interaction with `OnChargeTransaction`.
+pub(crate) type OnChargeTransactionOf<T> =
+    <T as pallet_transaction_payment::Config>::OnChargeTransaction;
+// Balance type alias.
+pub(crate) type BalanceOf<T> = <OnChargeTransactionOf<T> as OnChargeTransaction<T>>::Balance;
+
+// Type alias used for interaction with fungibles (assets).
+// Balance type alias.
+pub(crate) type AssetBalanceOf<T> =
+    <<T as pallet_asset_tx_payment::Config>::Fungibles as Inspect<
+        <T as frame_system::Config>::AccountId,
+    >>::Balance;
+
+// Type aliases used for interaction with `OnChargeAssetTransaction`.
+// Balance type alias.
+pub(crate) type ChargeAssetBalanceOf<T> =
+    <<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::Balance;
+// Asset id type alias.
+pub(crate) type ChargeAssetIdOf<T> =
+    <<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::AssetId;
+// Liquity info type alias.
+pub(crate) type ChargeAssetLiquidityOf<T> =
+    <<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::LiquidityInfo;
+
+// Type to track whether call is ISMP call or not
+pub(crate) type IsIsmpCall = bool;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::config]
+    pub trait Config:
+        frame_system::Config + pallet_ismp::Config + pallet_asset_tx_payment::pallet::Config
+    {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// The fungibles instance used to pay for transactions in assets.
+        // type Fungibles: Balanced<Self::AccountId>;
+        /// The actual transaction charging logic that charges the fees.
+        type OnChargeAssetTransaction: OnChargeAssetTransaction<Self>;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// A transaction fee `actual_fee`, of which `tip` was added to the minimum inclusion fee,
+        /// has been paid by `who` in an asset `asset_id`.
+        AssetTxFeePaid {
+            who: T::AccountId,
+            actual_fee: AssetBalanceOf<T>,
+            tip: AssetBalanceOf<T>,
+            asset_id: Option<ChargeAssetIdOf<T>>,
+        },
+    }
+}
+
+/// Require the transactor pay for themselves and maybe include a tip to gain additional priority
+/// in the queue. Allows paying via both `Currency` as well as `fungibles::Balanced`.
+///
+/// Wraps the transaction logic in [`pallet_transaction_payment`] and extends it with assets.
+/// An asset id of `None` falls back to the underlying transaction payment via the native currency.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct ChargeAssetTxPayment<T: Config> {
```

Comment:
just create a wrapper around the ChargeAssetTxPayment already defined in the pallet-asset-tx-payment, This task should only have an implementation for signed extension and nothing else, just one file.